File: 143. Reorder List/143. Reorder List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode *right = head, *left = head, *prev = nullptr, *tmp;
        while(right && right->next){
            right = right->next->next;
            left = left->next;
        }
        prev = left;
        left = left->next;
        prev->next = nullptr;
        while(left){
            tmp = left->next;
            left->next = prev;
            prev = left;
            left = tmp;
        }
        
        left = head;
        right = prev;
        while(left != right && left->next && right->next){
            tmp = right->next;
            right->next = left->next;
            left->next = right;
            left = right->next;
            right = tmp;
        }
        return;
    }
};

File: 143. Reorder List/143. Reorder List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        ListNode left = head, right = head, prev;
        while(right != null && right.next != null){
            left = left.next;
            right = right.next.next;
        }
        prev = left;
        left = left.next;
        prev.next = null;
        while(left != null){
            right = left.next;
            left.next = prev;
            prev = left;
            left = right;
        }
        right = head;
        left = prev;
        while(left.next != null){
            prev = right.next;
            right.next = left;
            left = prev;
            right = right.next;
        }
    }
}